api/client/client.go

```diff
@@ -1319,6 +1319,55 @@ func (c *Client) DeleteAllNodes(ctx context.Context, namespace string) error {
 	return trail.FromGRPC(err)
 }
 
+// StreamSessionEvents streams audit events from a given session recording.
+func (c *Client) StreamSessionEvents(ctx context.Context, sessionID string, startIndex int) (chan events.AuditEvent, chan error) {
+	request := &proto.StreamSessionEventsRequest{
+		SessionID:  sessionID,
+		StartIndex: int32(startIndex),
+	}
+
+	ch := make(chan events.AuditEvent)
+
+	stream, err := c.grpc.StreamSessionEvents(ctx, request)
+	if err != nil {
+		e := make(chan error)
+		go func() {
+			e <- trace.Wrap(err)
+			close(ch)
+		}()
+
+		return ch, e
+	}
+
+	e := make(chan error)
+
+	go func() {
+	outer:
+		for {
+			oneOf, err := stream.Recv()
+			if err != nil {
+				if err != io.EOF {
+					e <- trace.Wrap(trail.FromGRPC(err))
+				}
+
+				break outer
+			}
+
+			event, err := events.FromOneOf(*oneOf)
+			if err != nil {
+				e <- trace.Wrap(trail.FromGRPC(err))
+				break outer
+			}
+
+			ch <- event
+		}
+
+		close(ch)
```

Comment:
Looking at the code organization I think it's theoretically possible to close the channel twice? If we signaling the error using a second channel, do we need to close the channel here and above risking panic?

Reply:
I'm not sure how we could close it twice. Here only the sender (this bit of code) closes it and it only does so either on error or when it runs out of events, both of which make sense. We could say that we don't close ch on error but I am not sure what benefits it provides.